incubator/binding-kafka-grpc/src/main/java/io/aklivity/zilla/runtime/binding/kafka/grpc/internal/stream/KafkaGrpcRemoteServerFactory.java

```diff
@@ -0,0 +1,1994 @@
+            int offset,
+            int length)
+        {
+            doData(grpc, originId, routedId, initialId, initialSeq, initialAck, initialMax,
+                traceId, authorization, budgetId, reserved, flags, buffer, offset, length, emptyRO);
+
+            initialSeq += reserved;
+            remoteServer.fetch.replyReserved -= length;
+
+            assert initialSeq <= initialAck + initialMax;
+        }
+
+        private void doGrpcAbort(
+            long traceId,
+            long authorization)
+        {
+            if (KafkaGrpcState.replyOpened(state) && !KafkaGrpcState.replyClosed(state))
+            {
+                final GrpcAbortExFW grpcAbortEx =
+                    grpcAbortExRW.wrap(extBuffer, 0, extBuffer.capacity())
+                        .typeId(grpcTypeId)
+                        .status(HEADER_VALUE_GRPC_INTERNAL_ERROR)
+                        .build();
+
+                doAbort(grpc, originId, routedId, initialId, initialSeq, initialAck, initialMax,
+                    traceId, authorization, grpcAbortEx);
+            }
+            state = KafkaGrpcState.closeInitial(state);
+        }
+
+        private void doGrpcEnd(
+            long traceId,
+            long authorization)
+        {
+            if (!KafkaGrpcState.replyClosed(state))
+            {
+                state = KafkaGrpcState.closeReply(state);
+
+                doEnd(grpc, originId, routedId, initialId, initialSeq, initialAck, initialMax,
+                    traceId, authorization);
+            }
+        }
+
+        private void doGrpcWindow(
+            long authorization,
+            long traceId,
+            long budgetId,
+            int padding,
+            int capabilities)
+        {
+            replyAck = producer.initialAck;
+            replyMax = producer.initialMax;
+
+            doWindow(grpc, originId, routedId, replyId, replySeq, replyAck, replyMax,
+                traceId, authorization, budgetId, padding, capabilities);
+        }
+
+        private void doGrpcReset(
+            long traceId,
+            long authorization)
+        {
+            if (KafkaGrpcState.replyOpening(state) &&
+                !KafkaGrpcState.replyClosed(state))
+            {
+                state = KafkaGrpcState.closeReply(state);
+
+                final GrpcResetExFW grpcResetEx =
+                    grpcResetExRW.wrap(extBuffer, 0, extBuffer.capacity())
+                        .typeId(grpcTypeId)
+                        .status(HEADER_VALUE_GRPC_INTERNAL_ERROR)
+                        .build();
+
+                doReset(grpc, originId, routedId, replyId, replySeq, replyAck, replyMax,
+                        traceId, authorization, grpcResetEx);
+            }
+        }
+
+        private void onGrpcClientClosed()
+        {
+            if (KafkaGrpcState.closed(state))
+            {
+                remoteServer.grpcClients.remove(correlationId);
+            }
+        }
```

Comment:
This feels like it should be handled in a callback on the `RemoteServer` instead.

Perhaps named `removeIfClosed(correlationId)`.

Reply:
Is this what you mean by moving it to RemoteServer we are going to call both get and remove right